rfcs/0049-flakes.md

```diff
@@ -0,0 +1,709 @@
+  configurations.
+
+* Nix projects lack discoverability and a standard structure. For
+  example, it's just convention that a repository has a `release.nix`
+  for Hydra jobs and a `default.nix` for packages.
+
+* There is no standard way to compose Nix projects. Typical ways are
+  to rely on the Nix search path (e.g. `import <nixpkgs>`) or to use
+  `fetchGit` or `fetchTarball`. The former has poor reproducibility,
+  while the latter is bad UX because of the need to manually update
+  Git hashes to update dependencies.
+
+* `nix-channel` needs a replacement: channels are hard to create,
+  users cannot easily pin specific versions of channels, channels
+  interact in *ad hoc* ways with the Nix search path, and so on.
+
+The flakes mechanism seeks to address all these problems. It can be
+seen as a "Cargo/NPM/... for Nix" except that it's built into Nix
+rather than a separate tool.
+
+# Detailed design
+[design]: #detailed-design
+
+## Flakes
+
+A flake is a Git repository that contains a file named `flake.nix` in
+the root directory. (In the future, there might be other types of
+flakes, such as Mercurial repositories or tarballs.) `flake.nix`
+specifies some metadata about the flake such as dependencies (called
+*inputs*), as well as its *outputs* (the Nix values such as packages
+or NixOS modules provided by the flake).
+
+As an example, below is the `flake.nix` of
+[`dwarffs`](https://github.com/edolstra/dwarffs) (a FUSE filesystem
+for automatically fetching DWARF debug symbols by ELF build ID). It
+depends on the Nixpkgs flake and provides a package (i.e. an
+installable derivation) and a NixOS module.
+
+```
+{
+  name = "dwarffs";
+
+  edition = 201906;
+
+  description = "A filesystem that fetches DWARF debug info from the Internet on demand";
+
+  inputs = [ "nixpkgs" ];
+
+  outputs = inputs: rec {
+    packages.dwarffs =
+      with inputs.nixpkgs.packages;
+      with inputs.nixpkgs.builders;
+      with inputs.nixpkgs.lib;
+
+      stdenv.mkDerivation {
+        name = "dwarffs-0.1.${substring 0 8 inputs.self.lastModified}";
+
+        buildInputs = [ fuse nix nlohmann_json boost ];
+
+        NIX_CFLAGS_COMPILE = "-I ${nix.dev}/include/nix -include ${nix.dev}/include/nix/config.h -D_FILE_OFFSET_BITS=64";
+
+        src = inputs.self;
+
+        installPhase =
+          ''
+            mkdir -p $out/bin $out/lib/systemd/system
+
+            cp dwarffs $out/bin/
+            ln -s dwarffs $out/bin/mount.fuse.dwarffs
+
+            cp ${./run-dwarffs.mount} $out/lib/systemd/system/run-dwarffs.mount
+            cp ${./run-dwarffs.automount} $out/lib/systemd/system/run-dwarffs.automount
+          '';
+      };
+
+    nixosModules.dwarffs = import ./module.nix inputs;
+
+    defaultPackage = packages.dwarffs;
+
+    checks.build = packages.dwarffs;
+  };
+}
+```
+
+A flake has the following attributes:
+
+* `edition`: A number that specifies the version of the flake
```

Comment:
(nitpick) while `edition` is a valid term, it's somewhat vague to me. It's common to use `revision` or `version`.

Reply:
I think this is inspired by https://doc.rust-lang.org/nightly/edition-guide/editions/index.html

Reply:
But it has a very different meaning. In rust edition is opt-in: `When a new edition becomes available in the compiler, crates must explicitly opt in to it to take full advantage`.

According to text below it seems Nix will force the user to upgrade the version, so it's a different meaning.

Reply:
No, the intent is definitely that Nix would support previous editions (though maybe not forever). You only get an error if you use a flake that uses an edition more recent than what your version of Nix supports.

Reply:
I think this should be optional for the first edition, since it's just line noise if we end up never breaking compatibility.